src/transmission.js

```diff
@@ -34,6 +35,54 @@ const eachPromise = (arr, iteratorFn) =>
         });
     }, Promise.resolve());
 
+const partition = (arr, keyfn, createfn, addfn) => {
+  let result = Object.create(null);
+  arr.forEach((v) => {
+    let key = keyfn(v);
+    if (!result[key]) {
+      result[key] = createfn(v);
+    } else {
+      addfn(result[key], v);
+    }
+  });
+  return result;
+};
+
+class BatchEndpointAggregator {
+  constructor(events) {
+    this.batches = partition(events,
+                             /* keyfn */
+                             (ev) => `${ev.apiHost}_${ev.writeKey}_${ev.dataset}`,
+                             /* createfn */
+                             (ev) => ({
+                               apiHost: ev.apiHost,
+                               writeKey: ev.writeKey,
+                               dataset: ev.dataset,
+                               events: [ev]
+                             }),
+                             /* addfn */
+                             (batch, ev) => batch.events.push(ev));
+  }
+
+  encodeBatchEvents (events) {
+    let numEncoded = 0;
+    let encoded = "[" +
+          events.map((ev, i) => {
+            let evEncoded;
+            try {
+              evEncoded = JSON.stringify(ev);
+              numEncoded ++;
+            } catch (e) {
+              ev.encodeError = e;
+              evEncoded = "null";
```

Comment:
in the face of encoding Errors in other SDKs, do we also just send a handful of `null`s up to the server and deal with them on return? if something fails to encode, how painful would it be to just yank it out of the batch in the first place?